### PR TITLE
Add ExportableDataPayload::eq_dyn

### PR DESCRIPTION
**Pull request by sffc**
Toward https://github.com/unicode-org/icu4x/issues/2683

### CODE REVIEW

File: provider/core/src/datagen/payload.rs [hunk 1]
```diff
@@ -143,3 +169,28 @@ pub struct ExportMarker {}
...
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::hello_world::*;
+
+    #[test]
+    fn test_compare_with_dyn() {
+        let payload1: DataPayload<HelloWorldV1Marker> = DataPayload::from_owned(HelloWorldV1 {
```
Comment (robertbastian) [inline provider/core/src/datagen/payload.rs:180]: What's the plan for comparing `DataPayload<ExportMarker>`, i.e. the upcasted one? Are you planning to do that in a separate PR?

**Comment by sffc** [comment]
The build failure is because I need to add more PartialEq impls. Good news is it seems I need only these 6:

```
error[E0277]: can't compare `icu_list::provider::ListFormatterPatternsV1<'a>` with `icu_list::provider::ListFormatterPatternsV1<'a>`

error[E0277]: can't compare `icu_properties::provider::names::PropertyValueNameToEnumMapV1<'a>` with `icu_properties::provider::names::PropertyValueNameToEnumMapV1<'a>`

error[E0277]: can't compare `icu_properties::provider::names::PropertyEnumToValueNameLinearMapV1<'a>` with `icu_properties::provider::names::PropertyEnumToValueNameLinearMapV1<'a>`

error[E0277]: can't compare `icu_properties::provider::names::PropertyEnumToValueNameSparseMapV1<'a>` with `icu_properties::provider::names::PropertyEnumToValueNameSparseMapV1<'a>`

error[E0277]: can't compare `icu_properties::provider::names::PropertyEnumToValueNameLinearTiny4MapV1<'a>` with `icu_properties::provider::names::PropertyEnumToValueNameLinearTiny4MapV1<'a>`

error[E0277]: can't compare `WeekDataV1` with `WeekDataV1`
```

**Review by Manishearth — DISMISSED** [review]

**Review by robertbastian — DISMISSED** [review]

**Review by Manishearth — DISMISSED** [review]

**Review by Manishearth — APPROVED** [review]
